File: tools/crossify_static_server.py
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
import os


ROOT = Path(__file__).resolve().parents[1]
PUBLIC = ROOT / "public"


class CrossifyHandler(SimpleHTTPRequestHandler):
    def translate_path(self, path):
        path_only = path.split("?", 1)[0].split("#", 1)[0]
        if path_only == "/" or path_only == "":
            return str(ROOT / "index.html")
        if path_only.startswith("/assets/"):
            return str(PUBLIC / path_only.lstrip("/"))
        return str(ROOT / path_only.lstrip("/"))

    def end_headers(self):
        self.send_header("Cache-Control", "no-store")
        super().end_headers()


if __name__ == "__main__":
    os.chdir(ROOT)
    server = ThreadingHTTPServer(("127.0.0.1", 5173), CrossifyHandler)
    print("Crossify local server: http://127.0.0.1:5173", flush=True)
    server.serve_forever()

File: tools/crossify_static_server.mjs
import { createReadStream, statSync } from 'node:fs'
import { createServer } from 'node:http'
import { extname, join, normalize, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

const ROOT = resolve(fileURLToPath(new URL('..', import.meta.url)))
const PUBLIC = join(ROOT, 'public')
const PORT = 5173

const TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.json': 'application/json; charset=utf-8',
  '.xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
}

function inside(base, file) {
  const relative = normalize(file).toLowerCase()
  return relative.startsWith(normalize(base).toLowerCase())
}

function fileForUrl(url) {
  const parsed = new URL(url, `http://127.0.0.1:${PORT}`)
  const path = decodeURIComponent(parsed.pathname)
  if (path === '/') return join(ROOT, 'index.html')
  if (path.startsWith('/assets/')) return join(PUBLIC, path.slice(1))
  return join(ROOT, path.slice(1))
}

createServer((req, res) => {
  try {
    const file = resolve(fileForUrl(req.url || '/'))
    if (!inside(ROOT, file) && !inside(PUBLIC, file)) {
      res.writeHead(403).end('Forbidden')
      return
    }
    const st = statSync(file)
    if (!st.isFile()) {
      res.writeHead(404).end('Not found')
      return
    }
    res.writeHead(200, {
      'Content-Type': TYPES[extname(file).toLowerCase()] || 'application/octet-stream',
      'Content-Length': st.size,
      'Cache-Control': 'no-store',
    })
    createReadStream(file).pipe(res)
  } catch {
    res.writeHead(404).end('Not found')
  }
}).listen(PORT, '127.0.0.1')
